src/statement.js

```diff
@@ -114,22 +114,23 @@ pp.parseStatement = function(declaration, topLevel, exports) {
         this.raise(this.start, "'import' and 'export' may appear only with 'sourceType: module'")
     }
     return starttype === tt._import ? this.parseImport(node) : this.parseExport(node, exports)
-  }
 
-  if (this.isAsyncFunction() && declaration) {
-    this.next()
-    return this.parseFunctionStatement(node, true)
-  }
+    // If the statement does not start with a statement keyword or a
+    // brace, it's an ExpressionStatement or LabeledStatement. We
+    // simply start parsing an expression, and afterwards, if the
+    // next token is a colon and the expression was a simple
+    // Identifier node, we switch to interpreting it as a label.
+  default:
+    if (this.isAsyncFunction() && declaration) {
+      this.next()
+      return this.parseFunctionStatement(node, true)
+    }
 
-  // If the statement does not start with a statement keyword or a
-  // brace, it's an ExpressionStatement or LabeledStatement. We
-  // simply start parsing an expression, and afterwards, if the
-  // next token is a colon and the expression was a simple
-  // Identifier node, we switch to interpreting it as a label.
-  let maybeName = this.value, expr = this.parseExpression()
-  if (starttype === tt.name && expr.type === "Identifier" && this.eat(tt.colon))
-    return this.parseLabeledStatement(node, maybeName, expr)
-  else return this.parseExpressionStatement(node, expr)
+    let maybeName = this.value, expr = this.parseExpression()
+    if (starttype === tt.name && expr.type === "Identifier" && this.eat(tt.colon))
+      return this.parseLabeledStatement(node, maybeName, expr)
+    else return this.parseExpressionStatement(node, expr)
+  }
```

Comment:
Browsers seem to error on any function-inside-if in strict mode, so I guess you're right, and reverting the old patch entirely makes sense.